src/scatter.h

```diff
@@ -33,222 +35,335 @@ template <class T = float> class FormFactorSphere {
+template <typename T = float, typename TSamplingPolicy = SamplingPolicy<T>> class StructureFactorPBC : private TSamplingPolicy {
+    //! sample directions (h,k,l)
+    const std::vector<Point> directions = {
         {1, 0, 0}, {0, 1, 0},  {0, 0, 1},                                      // 3 permutations
         {1, 1, 0}, {0, 1, 1},  {1, 0, 1},  {-1, 1, 0}, {-1, 0, 1}, {0, -1, 1}, // 6 permutations
         {1, 1, 1}, {-1, 1, 1}, {1, -1, 1}, {1, 1, -1}                          // 4 permutations
     };
 
-  public:
-    StructureFactor(int number_of_scalings) : pmax(number_of_scalings) {}
+    const int p_max;  //!< multiples of q to be sampled
+    using TSamplingPolicy::addSampling;
 
-    int pmax; //!< Multiples of q to be sampled
+  public:
+    StructureFactorPBC(int q_multiplier) : p_max(q_multiplier){}
 
-    template <class Tpvec> void sample(const Tpvec &positions, double boxlength) {
-        auto func = [&](Point &dir) {
-            for (int p = 1; p <= pmax; p++) {                  // loop over multiples of q
-                Point _q = (2 * pc::pi * p / boxlength) * dir; // scattering vector
-                double sum_sin = 0, sum_cos = 0;               // temporary sums
-                for (auto &r : positions) {                    // loop over positions
-                    double qr = _q.dot(r);                     // scalar product q*r
+    template <class Tpvec> void sample(const Tpvec &positions, const double boxlength) {
+        // https://gcc.gnu.org/gcc-9/porting_to.html#ompdatasharing
+        // #pragma omp parallel for collapse(2) default(none) shared(directions, p_max, boxlength) shared(positions)
+        #pragma omp parallel for collapse(2) default(shared)
+        for (int i = 0; i < directions.size(); ++i) {
+            for (int p = 1; p <= p_max; ++p) {                                // loop over multiples of q
+                const Point q = (2 * pc::pi * p / boxlength) * directions[i]; // scattering vector
+                #ifdef __GNUC__
+                // When sine and cosine is computed in separate loops, advanced sine and cosine implementation
+                // utilizing SIMD instructions may be used to get at least 4 times performance boost.
+                // As of January 2020, only GCC exploits this using libmvec library if --ffast-math is enabled.
+                std::vector<T> qr_products(positions.size());
+                std::transform(positions.begin(), positions.end(), qr_products.begin(),
+                               [&q](auto &r) { return q.dot(r); });
+                // as of January 2020 the std::transform_reduce is not implemented in libc++
+                T sum_sin = 0;
+                for (auto &qr : qr_products) {
+                    sum_sin += std::sin(qr);
+                }
+                // as of January 2020 the std::transform_reduce is not implemented in libc++
+                T sum_cos = 0;
+                for (auto &qr : qr_products) {
```

Comment:
How well does Eigen fare for this operation? It's supposed to handle vectorisation on a number of architectures. I believe the operation here would by `qr_products.sin().sum()`. Note also that for the coming Eigen 4, iterators are supported.

Reply:
Btw. your sin/cos optimization may be applicable also in `energy.h:127` for Ewald summation.

Reply:
Update: on GCC9/MacOS, the separate cos/sin (ifdef _GNUC_) are slower

Reply:
The following gives a factor four speed-up on GCC and Clang:
~~~ cpp
Eigen::VectorXd qr_products;
qr_products.noalias() = positions * q;
T sum_sin = qr_products.array().sin().sum();
T sum_cos = qr_products.array().cos().sum();
~~~
where `positions` must by a `MatrixXd(N,3)` object. Best to await a unit-test to see if it's doing the right thing, though :-)